File: src/main/java/cz/solutions/cockroach/SalesReportPreparation.java
package cz.solutions.cockroach;

import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class SalesReportPreparation {
    private static final DateTimeFormatter DATE_FORMATTERTER = DateTimeFormat.forPattern("dd.MM.YYYY").withZoneUTC();

    public SalesReport generateSalesReport(List<SaleRecord> salesRecords, DateInterval interval, ExchangeRateProvider exchangeRateProvider) {

        List<SaleRecord> saleRecords = salesRecords.stream()
                .filter(a -> interval.contains(a.getDate()))
                .sorted(Comparator.comparing(SaleRecord::getDate))
                .toList();

        List<PrintableSale> printableSalesList = new ArrayList<>();
        double sellDollarValue = 0;
        double sellCroneValue = 0;
        double profitDolarValue = 0;
        double recentProfitCroneValue = 0;
        double recentSellCroneValue = 0;
        double recentBuyCroneValue = 0;

        double totalAmount = 0;

        for (SaleRecord sale : saleRecords) {
            double exchange = exchangeRateProvider.rateAt(sale.getDate());

            double partialsellDolarValue = sale.getQuantity() * sale.getSalePrice();
            double partialsellCroneValue = partialsellDolarValue * exchange;
            double buyPriceDolarValue = sale.getQuantity() * sale.getPurchaseFmv();

            double partialProfitValue = partialsellDolarValue - buyPriceDolarValue;
            double partialRecentProfitValue = sale.isTaxable() ? partialProfitValue : 0;
            double partialRecentProfitCroneValue = partialRecentProfitValue * exchange;

            sellDollarValue += partialsellDolarValue;
            sellCroneValue += partialsellCroneValue;

            profitDolarValue += partialProfitValue;
            recentProfitCroneValue += partialRecentProfitCroneValue;

            if(sale.isTaxable()){
                recentSellCroneValue += partialsellCroneValue;
                recentBuyCroneValue += buyPriceDolarValue * exchange;
            }
            totalAmount += sale.getQuantity();


            printableSalesList.add(
                    new PrintableSale(
                            DATE_FORMATTERTER.print(sale.getDate()),
                            DATE_FORMATTERTER.print(sale.getPurchaseDate()),
                            FormatingHelper.formatDouble(sale.getQuantity()),
                            FormatingHelper.formatExchangeRate(exchange),

                            FormatingHelper.formatDouble(sale.getPurchaseFmv()),
                            FormatingHelper.formatDouble(sale.getSalePrice()),
                            FormatingHelper.formatDouble(sale.getSalePrice() - sale.getPurchaseFmv()),

                            FormatingHelper.formatDouble(partialsellDolarValue),
                            FormatingHelper.formatDouble(partialProfitValue),
                            FormatingHelper.formatDouble(partialsellCroneValue),
                            FormatingHelper.formatDouble(partialRecentProfitCroneValue)
                    )
            );
        }

        double profitForTax;
        double sellCroneForTax;
        double buyCroneForTax;
        if (sellCroneValue < 100000) {
            // no need to pay taxes
            profitForTax = 0;
            sellCroneForTax = 0;
            buyCroneForTax=0;
        } else {
            //pay taxes only from items bought in last 3 years
            profitForTax = recentProfitCroneValue;
            sellCroneForTax = recentSellCroneValue;
            buyCroneForTax = recentBuyCroneValue;
        }
        return new SalesReport(
                printableSalesList,
                sellCroneValue,
                sellDollarValue,
                profitDolarValue,
                recentProfitCroneValue,
                totalAmount,
                profitForTax,
                sellCroneForTax,
                buyCroneForTax
        );


    }
}

File: src/main/kotlin/cz/solutions/cockroach/SalesReportPreparation.kt
package cz.solutions.cockroach

import org.joda.time.format.DateTimeFormat
import org.joda.time.format.DateTimeFormatter

object SalesReportPreparation {
    private val DATE_FORMATTER: DateTimeFormatter = DateTimeFormat.forPattern("dd.MM.YYYY").withZoneUTC()

    fun generateSalesReport(salesRecords: List<SaleRecord>, interval: DateInterval, exchangeRateProvider: ExchangeRateProvider): SalesReport {

        val filteredSaleRecords = salesRecords
            .filter { interval.contains(it.date) }
            .sortedBy { it.date }

        var sellDollarValue = 0.0
        var sellCroneValue = 0.0
        var buyCroneValue = 0.0
        var buyDollarValue = 0.0
        var profitCroneValue = 0.0
        var recentProfitCroneValue = 0.0
        var recentSellCroneValue = 0.0
        var recentBuyCroneValue = 0.0
        var totalAmount = 0.0

        val printableSalesList = filteredSaleRecords.map { sale ->
            val exchange = exchangeRateProvider.rateAt(sale.date)

            val partialSellDolarValue = sale.quantity * sale.salePrice
            val partialSellCroneValue = partialSellDolarValue * exchange

            val buyPriceDolarValue = sale.quantity * sale.purchaseFmv
            val buyPriceCroneValue = buyPriceDolarValue * exchange //todo

            val partialProfitCroneValue = partialSellCroneValue - buyPriceCroneValue
            val partialRecentProfitCroneValue = if (sale.isTaxable()) partialProfitCroneValue else 0.0

            val (partialRecentBuyCroneValue,partialRecentSellCroneValue) =if (sale.isTaxable()){
                listOf(buyPriceCroneValue,partialSellCroneValue)
            }else{
                listOf(0.0,0.0)
            }


            buyDollarValue += buyPriceDolarValue
            buyCroneValue+=buyPriceCroneValue
            sellDollarValue += partialSellDolarValue
            sellCroneValue += partialSellCroneValue
            profitCroneValue += partialProfitCroneValue
            recentProfitCroneValue += partialRecentProfitCroneValue

            recentSellCroneValue += partialRecentSellCroneValue
            recentBuyCroneValue += partialRecentBuyCroneValue

            totalAmount += sale.quantity

            PrintableSale(
                amount = FormatingHelper.formatDouble(sale.quantity),

                purchaseDate = DATE_FORMATTER.print(sale.purchaseDate),
                onePurchaseDollar = FormatingHelper.formatDouble(sale.purchaseFmv),
                purchaseDollar = FormatingHelper.formatDouble(buyPriceDolarValue),
                purchaseExchange = FormatingHelper.formatExchangeRate(exchange), //todo
                purchaseCrone = FormatingHelper.formatDouble(buyPriceCroneValue),
                recentPurchaseCrone =  FormatingHelper.formatDouble(partialRecentBuyCroneValue),

                date = DATE_FORMATTER.print(sale.date),
                oneSellDollar = FormatingHelper.formatDouble(sale.salePrice),
                sellDolar = FormatingHelper.formatDouble(partialSellDolarValue),
                sellExchange = FormatingHelper.formatExchangeRate(exchange),
                sellCrone = FormatingHelper.formatDouble(partialSellCroneValue),
                recentSellCrone = FormatingHelper.formatDouble(partialRecentSellCroneValue),


                sellProfitCrone = FormatingHelper.formatDouble(partialProfitCroneValue),
                sellRecentProfitCrone = FormatingHelper.formatDouble(partialRecentProfitCroneValue)
            )
        }

        val (profitForTax, sellCroneForTax, buyCroneForTax) = if (sellCroneValue < 100000) {
            Triple(0.0, 0.0, 0.0)
        } else {
            Triple(recentProfitCroneValue, recentSellCroneValue, recentBuyCroneValue)
        }

        return SalesReport(
            printableSalesList = printableSalesList,
            sellCroneValue = sellCroneValue,
            recentSellCroneValue = recentSellCroneValue,
            sellDollarValue = sellDollarValue,
            profitCroneValue = profitCroneValue,
            recentProfitCroneValue = recentProfitCroneValue,
            totalAmount = totalAmount,
            profitForTax = profitForTax,
            sellCroneForTax = sellCroneForTax,
            buyCroneForTax = buyCroneForTax,
            buyCroneValue = buyCroneValue,
            recentBuyCroneValue = recentBuyCroneValue,
            buyDollarValue = buyDollarValue
        )
    }
}